pyeudiw/trust/handler/direct_trust_sd_jwt_vc.py

```diff
@@ -0,0 +1,129 @@
+
+    def _get_jwks_by_reference(self, jwks_reference_uri: str) -> dict:
+        """
+        call the jwks endpoint if jwks is defined by reference
+        """
+        if self.cache_ttl:
+            resp = cacheable_get_http_url(self.cache_ttl, jwks_reference_uri, self.httpc_params, http_async=self.http_async_calls)
+        else:
+            resp = get_http_url([jwks_reference_uri], self.httpc_params, http_async=self.http_async_calls)[0]
+        return resp.json()
+
+    def _extract_jwks_from_jwk_metadata(self, metadata: dict) -> dict:
+        """
+        parse the jwk metadata document and return the jwks
+        NOTE: jwks might be in the document by value or by reference
+        """
+        jwks: dict[Literal["keys"], list[dict]] | None = metadata.get("jwks", None)
+        jwks_uri: str | None = metadata.get("jwks_uri", None)
+        if (not jwks) and (not jwks_uri):
+            raise InvalidJwkMetadataException("invalid issuing key metadata: missing both claims [jwks] and [jwks_uri]")
+        if jwks:
+            # get jwks by value
+            return jwks
+        return self._get_jwks_by_reference(jwks_uri)
+
+    def build_issuer_jwk_endpoint(issuer_id: str, well_known_path_component: str) -> str:
+        baseurl = urlparse(issuer_id)
+        well_known_path = well_known_path_component + baseurl.path
+        well_known_url: str = ParseResult(baseurl.scheme, baseurl.netloc, well_known_path, baseurl.params, baseurl.query, baseurl.fragment).geturl()
+        return well_known_url
+
+    def build_issuer_metadata_endpoint(issuer: str, metadata_path_component: str) -> str:
+        issuer_normalized = issuer if issuer[-1] != '/' else issuer[:-1]
+        return issuer_normalized + metadata_path_component
+    
+        
+    def extract(self, issuer: str, trust_source: TrustSourceData) -> TrustSourceData:
+        """
+        Fetches the public key of the issuer by querying a given endpoint.
+        Previous responses might or might not be cached based on the cache_ttl
+        parameter.
+
+        :returns: a list of jwk(s)
+        """
+        if not issuer:
+            raise ValueError("invalid issuer: cannot be empty value")
+        
+        try:
+            self.get_metadata(issuer, trust_source)
+
+            md = self._get_jwk_metadata(issuer)
+            if not issuer == (obt_issuer := md.get("issuer", None)):
+                raise InvalidJwkMetadataException(f"invalid jwk metadata: obtained issuer :{obt_issuer}, expected issuer: {issuer}")
+            jwks = self._extract_jwks_from_jwk_metadata(md)
+            jwk_l: list[dict] = jwks.get("keys", [])
+            if not jwk_l:
+                raise InvalidJwkMetadataException("unable to find jwks in issuer jwk metadata")
+            
+            trust_source.add_keys(jwk_l)
+        except Exception as e:
+            self._log_warning("Extracting JWK" ,f"Failed to extract jwks from issuer {issuer}: {e}")
+    
+        return trust_source
```

Comment:
Proposal: drop the method output if the purpose of the function is to mutate the corresponding receiving input parameter - unless you _really_ need to mutate AND return (to do method chaining like some multiple precision arithmetic libraries - GMP in C  math/big in Go, etc).
IMO the presence of output matching an input might obfuscate the fact that the input was mutated.

This is mostly a code style suggestion. Feel free to ignore if it doesn't fit your personal preference.

Reply:
Resolved in [33c0a54](https://github.com/italia/eudi-wallet-it-python/pull/284/commits/33c0a54145e85bc387e4b4420245fc4af56abac8)